[ComponentPort][TCSACR-406] Add a guide for ComponentPort (#1427)
Samsung/tizen-docs: docs/application/dotnet/guides/app-management/component-port.md

# Component Port

Tizen components of the component-based applications can communicate with each other using component ports. Components can send and receive serializable objects through component port communication.

The main feature of the `Tizen.Applications.ComponentBased.ComponentPort` class include:
- Managing component port
  You can set up the component ports to [send and receive requests](#port) between components of component-based application with the `Tizen.Applications.ComponentBased.ComponentPort` class.

- Using component task

>   You can set up the component tasks to [run tasks](#task) and wait for events from other components with the `Tizen.Applications.ComponentBased.ComponentTask` class.

Reviewer: All is well, but we need one line space between: 

6-7 
9-10 

By including one line space, the formatting will be as desired.

Author: Done

## Prerequisites

To enable your component to use the component port functionality:

1.  You need two components to communicate with each other through the component port.
2.  To use the methods and properties of the [Tizen.Applications.ComponentBased.ComponentPort](https://samsung.github.io/TizenFX/API9/api/Tizen.Applications.ComponentBased.ComponentPort.html) and [Tizen.Applications.ComponentBased.ComponentTask](https://samsung.github.io/TizenFX/API9/api/Tizen.Applications.ComponentBased.ComponentPort.html) classes, include the [Tizen.Applications.ComponentBased](https://samsung.github.io/TizenFX/API9/api/Tizen.Applications.ComponentBased.html) namespace in your component:

    ```
    using Tizen.Applications.ComponentBased;
    ```

<a name="local"></a>
## Managing component port

> To send a request from one component `ClientService.Tizen` to another `ServerService.Tizen` component use the [Tizen.Applications.ComponentBased.ComponentPort](https://samsung.github.io/TizenFX/API9/api/Tizen.Applications.ComponentBased.ComponentPort.html) class as follows:

Reviewer: comma missed
```suggestion
To send a request from one component `ClientService.Tizen` to another `ServerService.Tizen` component, use the [Tizen.Applications.ComponentBased.ComponentPort](https://samsung.github.io/TizenFX/API9/api/Tizen.Applications.ComponentBased.ComponentPort.html) class as follows:

Author: Done

                _port.RequestReceived += OnRequestReceived;
                _thread = new Thread(new ThreadStart(OnThread));
                _thread.Start();
                return true;
            }

            public override void OnStartCommand(AppControl appControl, bool restarted)
            {
                Log.Info("OnStartCommand()");
                base.OnStartCommand(appControl, restarted);
            }

            public override void OnDestroy()
            {
                Log.Info("OnDestroy()");
                base.OnDestroy();

                if (_port != null)
                {
                    _port.RequestReceived -= OnRequestReceived;
                    _port.Dispose();
                    _port = null;
                }
            }
        }
    }
    ```

2.  Set up the receiving thread of the component.

- To have the receiving thread of the component wait for incoming requests, call `WaitForEvent()` of the `Tizen.Applications.ComponentBased.ComponentPort` class.
- If `WaitForEvent()` is called in the main thread, then `WaitForEvent()` can not be called until the `Cancel()` is called. 

> - `WaitForEvent()` starts the main loop to wait for events from other components. To avoid blocking the main thread, it's recommended to use the `Tizen.Applications.ComponentBased.ComponentTask` class.

Reviewer: ```suggestion
- `WaitForEvent()` starts the main loop to wait for events from other components. To avoid blocking the main thread, it is recommended to use the `Tizen.Applications.ComponentBased.ComponentTask` class.

Author: Done

> - To handle the access control, call `AddPrivilege()` of the `Tizen.Applications.ComponentBased.ComponentPort` class. If a client doesn't have permission, the request is rejected with the permission denied error.

Reviewer: ```suggestion
- To handle the access control, call `AddPrivilege()` of the `Tizen.Applications.ComponentBased.ComponentPort` class. If a client does not have permission, the request is rejected with the permission denied error.

Author: Done

# Component Port

Tizen components of the component-based applications can communicate with each other using component ports. Components can send and receive serializable objects through component port communication.

The main feature of the `Tizen.Applications.ComponentBased.ComponentPort` class include:
- Managing component port

  You can set up the component ports to [send and receive requests](#port) between components of component-based application with the `Tizen.Applications.ComponentBased.ComponentPort` class.

- Using component task

  You can set up the component tasks to [run tasks](#task) and wait for events from other components with the `Tizen.Applications.ComponentBased.ComponentTask` class.

## Prerequisites

To enable your component to use the component port functionality:

> 1.  You need two components to communicate with each other through the component port.

Reviewer 2: @nawab-samsung, @hjhun, is this sequential? An ordered list is needed here?

Author: As far as I know, it's for readability and consistency.

            private void OnThread()
            {
                Log.Info("OnTrhead()");
                _port.WaitForEvent();
            }

            public override override OnCreate()
            {
                Log.Info("OnCreate()");
                base.OnCreate();
                _port = new ComponentPort("ServerService");
                _port.RequestReceived += OnRequestReceived;
                _thread = new Thread(new ThreadStart(OnThread));
                _thread.Start();
                return true;
            }

            public override void OnStartCommand(AppControl appControl, bool restarted)
            {
                Log.Info("OnStartCommand()");
                base.OnStartCommand(appControl, restarted);
            }

            public override void OnDestroy()
            {
                Log.Info("OnDestroy()");
                base.OnDestroy();

                if (_port != null)
                {
                    _port.RequestReceived -= OnRequestReceived;
                    _port.Dispose();
                    _port = null;
                }
            }
        }
    }
    ```

3.  Set up the receiving thread of the component.

>     1. To have the receiving thread of the component wait for incoming requests, call `WaitForEvent()` of the `Tizen.Applications.ComponentBased.ComponentPort` class.

Reviewer 2: @nawab-samsung, @hjhun, is this sequential? An ordered list is needed here?

Author: It's not sequential. The purpose of numbering is for readability.
If you think it's not needed, please tell me.

Reviewer 2: @hjhun, if these steps are not to be performed in an order (1.2.3) as listed, kindly change to unordered list

        ```csharp
        {
            _port = new ComponentPort("ServerService");
            _port.RequestReceived += OnRequestReceived;
            _port.AddPrivilege("http://tizen.org/privilege/datasharing");
            _thread = new Thread(new ThreadStart(OnThread));
            _thread.Start();
        }

        private void OnThread()
        {
            Log.Info("OnTrhead()");
            _port.WaitForEvent();
        }

        private void OnRequestedReceived(object sender, RequestEventArgs args)
        {
            Log.Info("Sender: + args.Sender);
            var request = args.Request;
            if (request.GetType() == typeof(Request))
            {
                var req = (Request)request;
                Log.Info("Command: " + req.Command);
                Log.Info("Sequence: " + req.Sequence);
                Log.Info("Message: " + req.Message);

                if (args.IsReplyRequested)
                {
                    args.Reply = new Response(req.Command, req.Sequence, 0);
                }
            }
        }
        ```

4.  Send the request in the sending component as follows:

> 

Reviewer 2: @nawab-samsung, @hjhun, is this sequential? An ordered list is needed here?

Author: Yes. I think it's needed.

# Component Port

Tizen components of the component-based applications can communicate with each other using component ports. Components can send and receive serializable objects through component port communication.

The main feature of the `Tizen.Applications.ComponentBased.ComponentPort` class include:
- Managing component port

  You can set up the component ports to [send and receive requests](#port) between components of component-based application with the `Tizen.Applications.ComponentBased.ComponentPort` class.

- Using component task

  You can set up the component tasks to [run tasks](#task) and wait for events from other components with the `Tizen.Applications.ComponentBased.ComponentTask` class.

## Prerequisites

To enable your component to use the component port functionality:

> 

Reviewer: Just a doubt, 
Dont we need privileges ?

To use classes and methods, the application has to request permission by adding the following privilege to the `tizen-manifest.xml` file:

Author: Privileges are not needed.
When creating the port, the own socket of the application is created.
The application can add privileges using AddPrivilege() for access control.

Reviewer: sure, thanks.

            }

            public int Sequence
            {
                get; set;
            }

            public int Result
            {
                get; set;
            }
        }

        public class ClientService : ServiceComponent
        {
            private static string LogTag = "ClientService";
            private static int _seq = 0;
            private ComponentPort _port;

            public override override OnCreate()
            {
                Log.Info("OnCreate()");
                base.OnCreate();
                _port = new ComponentPort("ClientService");
                return true;
            }

            public override void OnStartCommand(AppControl appControl, bool restarted)
            {
                Log.Info("OnStartCommand()");
                base.OnStartCommand(appControl, restarted);
            }

            public override void OnDestroy()
            {
                Log.Info("OnDestroy()");
                base.OnDestroy();

                if (_port != null)
                {
                    _port.Dispose();
                    _port = null;
                }
            }
        }
    }
    ```

>     1. Create a component port instance for receiving component `ServerService.Tizen` as follows:

Reviewer: There is the numbering issue, Shouldn't this be 2? and left-aligned just below the 1. 
```suggestion
2. Create a component port instance for receiving component `ServerService.Tizen` as follows:

Author: Done

>     ```

Reviewer: Global: 
Please add language identifier.

Like 
```Csharp

Author: Could you show me an example?
And, I think it's not needed. Because, it's a guide of .NET  application.
The developer knows the program language is CSharp.

Reviewer: @hjhun  I understand, you are right, but thats part of style guide. 

https://github.com/Samsung/tizen-docs/blob/master/styleguide/style.md#inline-code-blocks-with-language-identifier

Author: I got it. Done.

Reviewer: Thats so nice of you. 
Thanks

                    }
                }
            }

            private void OnThread()
            {
                Log.Info("OnTrhead()");
                _port.WaitForEvent();
            }

            public override override OnCreate()
            {
                Log.Info("OnCreate()");
                base.OnCreate();
                _port = new ComponentPort("ServerService");
                _port.RequestReceived += OnRequestReceived;
                _thread = new Thread(new ThreadStart(OnThread));
                _thread.Start();
                return true;
            }

            public override void OnStartCommand(AppControl appControl, bool restarted)
            {
                Log.Info("OnStartCommand()");
                base.OnStartCommand(appControl, restarted);
            }

            public override void OnDestroy()
            {
                Log.Info("OnDestroy()");
                base.OnDestroy();

                if (_port != null)
                {
                    _port.RequestReceived -= OnRequestReceived;
                    _port.Dispose();
                    _port = null;
                }
            }
        }
    }
    ```

> 2.  Set up the receiving thread of the component.

Reviewer: ```suggestion
3.  Set up the receiving thread of the component.

- To have the receiving thread of the component wait for incoming requests, call `WaitForEvent()` of the `Tizen.Applications.ComponentBased.ComponentPort` class.
- If `WaitForEvent()` is called in the main thread, then `WaitForEvent()` can not be called until the `Cancel()` is called. 
- `WaitForEvent()` starts the main loop to wait for events from other components. To avoid blocking the main thread, it is recommended to use the `Tizen.Applications.ComponentBased.ComponentTask` class.
- To handle the access control, call `AddPrivilege()` of the `Tizen.Applications.ComponentBased.ComponentPort` class. If a client does not have permission, the request is rejected with the permission denied error.

> - To handle the received request, define and register an event handler for the `RequestReceived` event of the `Tizen.Applications.ComponentBased.ComponentPort` class as follows:

Reviewer: Alignment:

We need align 244-248 to little right such that it all comes under point 3 & add numbering for consistency.

```suggestion
    1. To have the receiving thread of the component wait for incoming requests, call `WaitForEvent()` of the `Tizen.Applications.ComponentBased.ComponentPort` class.
    2. If `WaitForEvent()` is called in the main thread, then `WaitForEvent()` can not be called until `Cancel()` is called. 
    3. `WaitForEvent()` starts the main loop to wait for events from other components. To avoid blocking the main thread, it is recommended to use the `Tizen.Applications.ComponentBased.ComponentTask` class.
    4. To handle the access control, call `AddPrivilege()` of the `Tizen.Applications.ComponentBased.ComponentPort` class. If a client does not have permission, the request is rejected with the permission denied error.
    5. To handle the received request, define and register an event handler for the `RequestReceived` event of the `Tizen.Applications.ComponentBased.ComponentPort` class as follows:

Author: Done

>     ```

Reviewer: Kindly align the code segment inline to point 5

Author: Done

    {
        _port = new ComponentPort("ServerService");
        _port.RequestReceived += OnRequestReceived;
        _port.AddPrivilege("http://tizen.org/privilege/datasharing");
        _thread = new Thread(new ThreadStart(OnThread));
        _thread.Start();
    }

    private void OnThread()
    {
        Log.Info("OnTrhead()");
        _port.WaitForEvent();
    }

    private void OnRequestedReceived(object sender, RequestEventArgs args)
    {
        Log.Info("Sender: + args.Sender);
        var request = args.Request;
        if (request.GetType() == typeof(Request))
        {
            var req = (Request)request;
            Log.Info("Command: " + req.Command);
            Log.Info("Sequence: " + req.Sequence);
            Log.Info("Message: " + req.Message);

            if (args.IsReplyRequested)
            {
                args.Reply = new Response(req.Command, req.Sequence, 0);
            }
        }
    }
    ```

> 3.  Send the request in the sending component as follows:

Reviewer: ```suggestion
4.  Send the request in the sending component as follows:

Author: Done

    1. Call `WaitForEvent()` of the `Tizen.Applications.ComponentBased.ComponentPort` class in the thread, if you want to receive the reply from `ServerService.Tizen`.
    2. Call `WaitForPort`() of the `Tizen.Applications.ComponentBased.ComponentPort` class to wait until the target port is ready.
    3. Use `Send()` of the `Tizen.Applications.ComponentBased.ComponentPort` class to send the request.
    4. Provide the request to be sent as an instance of the `ClientService.Tizen.Request` class as follows:

>     ```

Reviewer: Kindly align the code segment inline to point 4.

Author: Done

> # Component Port

Author: Done

> # Component Port

Reviewer: @hjhun I request you to kindly add the file to toc_all.md 

### Application Management
#### [Overview](/application/dotnet/guides/app-management/overview.md)
#### [Application Icons](/application/dotnet/guides/app-management/app-icons.md)
#### [Application Controls](/application/dotnet/guides/app-management/app-controls.md)
#### [Application Manager](/application/dotnet/guides/app-management/app-manager.md)
#### [Event Manager](/application/dotnet/guides/app-management/event.md)
#### [Component Manager](/application/dotnet/guides/app-management/component-manager.md)
**#### [Component Port](/application/dotnet/guides/app-management/component-port.md)**
#### [Package Manager](/application/dotnet/guides/app-management/package-manager.md)
#### [Widget Control](/application/dotnet/guides/app-management/widget-control.md)
#### [Message Port](/application/dotnet/guides/app-management/message-port.md)
#### [Data Control](/application/dotnet/guides/app-management/data-control.md)

> # Component Port

Author: Oops. I'm sorry. I've fixed it.

> Tizen components of component-based application can communicate with each other using component ports. Components can send and receive serializable objects through component port communication.

Author: Done

The main feature of the `Tizen.Applications.ComponentBased.ComponentPort` class include:

> - Managing a Component Port

Author: I've changed C and P to c and p.

>   You can set up Component ports to [send and receive requests](#port) between components of component-based application with the `Tizen.Applications.ComponentBased.ComponentPort` class.

Author: Done. It's changed to 'the component ports'. I think small alphabet is better than this.

- Using a Component Task

>   You can set up Component tasks to [run tasks](#task) to wait for events from other components with the `Tizen.Applications.ComponentBased.ComponentTask` class.

Author: Done

## Prerequisites

To enable your component to use the component port functionality:

> 1.  You need 2 components to communicate with each other through the component port.

Author: Done

> 2.  To use the methods and properties of the [Tizen.Applications.ComponentBased.ComponentPort](https://samsung.github.io/TizenFX/latest/api/Tizen.Applications.ComponentBased.ComponentPort.html) and [Tizen.Applications.ComponentBased.ComponentTask](https://samsung.github.io/TizenFX/latest/api/Tizen.Applications.ComponentBased.ComponentPort.html) classes, include the [Tizen.Applications.ComponentBased](https://samsung.github.io/TizenFX/latest/api/Tizen.Applications.ComponentBased.html) namespace in your component:

Reviewer: Link is breaking, please verify

Author: The reference(doxygen) is not updated not yet. 
This feature will be added after Tizen 6.5 M1 release.

Author: I've updated it.
https://samsung.github.io/TizenFX/API9/api/Tizen.Applications.ComponentBased.ComponentPort.html

The link should be "https://samsung.github.io/TizenFx/API9/api/...".

    ```
    using Tizen.Applications.ComponentBased;
    ```

<a name="local"></a>
## Managing a Component Port

> To send a request from one component (`ClientService.Tizen`) to another (`ServerService.Tizen`) using the [Tizen.Applications.ComponentBased.ComponentPort](https://samsung.github.io/TizenFX/latest/api/Tizen.Applications.ComponentBased.ComponentPort.html) class:

Author: Done

1.  Create a component port instance in each component.

>     In the sending component (`ClientService.Tizen`):

Author: Done

            }

            public int Sequence
            {
                get; set;
            }

            public int Result
            {
                get; set;
            }
        }

        public class ClientService : ServiceComponent
        {
            private static string LogTag = "ClientService";
            private static int _seq = 0;
            private ComponentPort _port;

            public override override OnCreate()
            {
                Log.Info("OnCreate()");
                base.OnCreate();
                _port = new ComponentPort("ClientService");
                return true;
            }

            public override void OnStartCommand(AppControl appControl, bool restarted)
            {
                Log.Info("OnStartCommand()");
                base.OnStartCommand(appControl, restarted);
            }

            public override void OnDestroy()
            {
                Log.Info("OnDestroy()");
                base.OnDestroy();

                if (_port != null)
                {
                    _port.Dispose();
                    _port = null;
                }
            }
        }
    }
    ```

>     In the receiving application (`ServerService.Tizen`):

Author: Done

            private void OnThread()
            {
                Log.Info("OnTrhead()");
                _port.WaitForEvent();
            }

            public override override OnCreate()
            {
                Log.Info("OnCreate()");
                base.OnCreate();
                _port = new ComponentPort("ServerService");
                _port.RequestReceived += OnRequestReceived;
                _thread = new Thread(new ThreadStart(OnThread));
                _thread.Start();
                return true;
            }

            public override void OnStartCommand(AppControl appControl, bool restarted)
            {
                Log.Info("OnStartCommand()");
                base.OnStartCommand(appControl, restarted);
            }

            public override void OnDestroy()
            {
                Log.Info("OnDestroy()");
                base.OnDestroy();

                if (_port != null)
                {
                    _port.RequestReceived -= OnRequestReceived;
                    _port.Dispose();
                    _port = null;
                }
            }
        }
    }
    ```

2.  Set up the receiving thread of the component.

>     To have the receiving thread of the component wait for incoming requests, call the `WaitForEvent()` method of the `Tizen.Applications.ComponentBased.ComponentPort` class. If the `WaitForEvent()` method is called in the main thread, the next line of calling the `WaitForEvent()` method will not be called until the `Cancel()` is called. The `WaitForEvent()` method starts the main loop to wait for events from other components. To avoid blocking the main thread, it's recommended to use the `Tizen.Applications.ComponentBased.ComponentTask` class.

Author: Done

>     To handle access control, call the `AddPrivilege()` method of the `Tizen.Applications.ComponentBased.ComponentPort` class. If a client doesn't have a permission, the request will be rejected with the permission denied error.

Author: Done

>     To handle the received request, define and the register an event handler for the `RequestReceived` event of the `Tizen.Applications.ComponentBased.ComponentPort` class.

Author: Done

    ```
    {
        _port = new ComponentPort("ServerService");
        _port.RequestReceived += OnRequestReceived;
        _port.AddPrivilege("http://tizen.org/privilege/datasharing");
        _thread = new Thread(new ThreadStart(OnThread));
        _thread.Start();
    }

    private void OnThread()
    {
        Log.Info("OnTrhead()");
        _port.WaitForEvent();
    }

    private void OnRequestedReceived(object sender, RequestEventArgs args)
    {
        Log.Info("Sender: + args.Sender);
        var request = args.Request;
        if (request.GetType() == typeof(Request))
        {
            var req = (Request)request;
            Log.Info("Command: " + req.Command);
            Log.Info("Sequence: " + req.Sequence);
            Log.Info("Message: " + req.Message);

            if (args.IsReplyRequested)
            {
                args.Reply = new Response(req.Command, req.Sequence, 0);
            }
        }
    }
    ```

> 3.  In the sending component, to send the request, follow the steps:

Author: Done

>     a. Call the `WaitForEvent()` method `Tizen.Applications.ComponentBased.ComponentPort` class in the thread if you want to receive the reply from the ServerService.

Reviewer: ```suggestion
    1. Call `WaitForEvent()` of the `Tizen.Applications.ComponentBased.ComponentPort` class in the thread, if you want to receive the reply from the **ServerService.**
```
Is ServiceService correct, I guess not, please verify.

Author: Done

>     b. Call the `WaitForPort`() method `Tizen.Applications.ComponentBased.ComponentPort` class to wait until the target port is ready.

Reviewer: ```suggestion
   2. Call `WaitForPort`() of the `Tizen.Applications.ComponentBased.ComponentPort` class to wait until the target port is ready.
```

Author: Done

>     c. Use `Send()` method `Tizen.Applications.ComponentBased.ComponentPort` class to send the request.

Reviewer: ```suggestion
    3. Use `Send()` of the `Tizen.Applications.ComponentBased.ComponentPort` class to send the request.

Author: Done

>     d. Provide the request to be sent as an instance of the `ClientService.Tizen.Request` class as shown in the below code:

Reviewer: ```suggestion
    4. Provide the request to be sent as an instance of the `ClientService.Tizen.Request` class as follows:

Author: Done

    ```
    {
        _port = new ComponentPort("ClientService");
        _port.RequestReceived += OnRequestReceived;
        _thread = new Thread(new ThreadStart(OnThread));
        _thread.Start();
        ...
        await ComponentPort.WaitForPort("ServerService");

        try
        {
            var request = new Request("TestCommand", _seq++, "TestMessage");
            _port.Send("ServerService", 5000, request);
        }
        catch
        {
            Log.Error("Failed to send request");
        }
        ...
    }

    private void OnThread()
    {
        Log.Info("OnTrhead()");
        _port.WaitForEvent();
    }

    private void OnRequestedReceived(object sender, RequestEventArgs args)
    {
        Log.Info("Sender: + args.Sender);
        var response = args.Request;
        if (response.GetType() == typeof(Response))
        {
            var req = (Response)request;
            Log.Info("Command: " + req.Command);
            Log.Info("Sequence: " + req.Sequence);
            Log.Info("Result: " + req.Result);
        }
    }
    ```

<a name="task"></a>

> ## Using a Component Task

Reviewer: Global:

Component Port
Component Task 

These are to be written as is, am I correct? then we need to maintain consistency all over this page. 

```suggestion 
## Use Component Task

Author: Done

> Using [Tizen.Applications.ComponentBased.ComponentTask](https://samsung.github.io/TizenFX/latest/api/Tizen.Applications.ComponentBased.ComponentTask.html) class, an component can run a task to wait for events from other component properly.

Reviewer: ```suggestion
Using [Tizen.Applications.ComponentBased.ComponentTask](https://samsung.github.io/TizenFX/latest/api/Tizen.Applications.ComponentBased.ComponentTask.html) class, a component can run a task to wait for events from other component properly.

Author: Done

1. Create the component task instance in the component:

> (`ServerService.Tizen`):

Reviewer: ```suggestion
.1. Create a component task instance for receiving component `ServerService.Tizen` as follows:

Author: Done

                    }
                }
            }

            public override override OnCreate()
            {
                Log.Info("OnCreate()");
                base.OnCreate();
                _port = new ComponentPort("ServerService");
                _port.RequestReceived += OnRequestReceived;
                _task = new ComponentTask(_port);
                _task.Start();
                return true;
            }

            public override void OnStartCommand(AppControl appControl, bool restarted)
            {
                Log.Info("OnStartCommand()");
                base.OnStartCommand(appControl, restarted);
            }

            public override void OnDestroy()
            {
                Log.Info("OnDestroy()");
                base.OnDestroy();

                if (_task != null)
                {
                    _task.Stop();
                    _task = null;
                }

                if (_port != null)
                {
                    _port.RequestReceived -= OnRequestReceived;
                    _port.Dispose();
                    _port = null;
                }
            }
        }
    }
    ```

> 2.  To run the task, call the `Start` method of the `Tizen.Applications.ComponentBased.ComponentTask` class:

Reviewer: ```suggestion
2. To run the task, call the `Start` method of the `Tizen.Applications.ComponentBased.ComponentTask` class as follows:

Author: Done

    ```
    _port = new ComponentPort("ServerService");
    _port.RequestReceived += OnRequestReceived;
    _task = new ComponentTask(_port);
    _task.Start();
    ```

> ## Related Information

Reviewer: ```suggestion
## Related information

Author: Done

* Dependencies

>   -   Tizen 6.5 and Higher

Reviewer: ```suggestion
- Dependencies
  -   Tizen 6.5 and Higher

Author: Done

# Component Port

Tizen components of the component-based applications can communicate with each other using component ports. Components can send and receive serializable objects through component port communication.

The main feature of the `Tizen.Applications.ComponentBased.ComponentPort` class include:

> - Managing a component port

Reviewer: Just for consistency across the pages suggesting this:
e.g application/dotnet/guides/app-management/package-manager/
application/dotnet/guides/app-management/event/

```suggestion
- Managing component port

Author: Done

  You can set up the component ports to [send and receive requests](#port) between components of component-based application with the `Tizen.Applications.ComponentBased.ComponentPort` class.

> - Use component task

Reviewer: Just for consistency across the pages suggesting this:
```suggestion
- Using component task

Author: Done

  You can set up the component tasks to [run tasks](#task) and wait for events from other components with the `Tizen.Applications.ComponentBased.ComponentTask` class.

## Prerequisites

To enable your component to use the component port functionality:

1.  You need two components to communicate with each other through the component port.
2.  To use the methods and properties of the [Tizen.Applications.ComponentBased.ComponentPort](https://samsung.github.io/TizenFX/API9/api/Tizen.Applications.ComponentBased.ComponentPort.html) and [Tizen.Applications.ComponentBased.ComponentTask](https://samsung.github.io/TizenFX/API9/api/Tizen.Applications.ComponentBased.ComponentPort.html) classes, include the [Tizen.Applications.ComponentBased](https://samsung.github.io/TizenFX/API9/api/Tizen.Applications.ComponentBased.html) namespace in your component:

    ```
    using Tizen.Applications.ComponentBased;
    ```

<a name="local"></a>

> ## Managing a component port

Author: Done

To send a request from one component `ClientService.Tizen` to another `ServerService.Tizen` component use the [Tizen.Applications.ComponentBased.ComponentPort](https://samsung.github.io/TizenFX/API9/api/Tizen.Applications.ComponentBased.ComponentPort.html) class as follows:

> 1.  Create a component port instance for sending component `ClientService.Tizen` as follows.

Author: Done

>     In the sending component (`ClientService.Tizen`):

Author: Done

            }

            public int Sequence
            {
                get; set;
            }

            public int Result
            {
                get; set;
            }
        }

        public class ClientService : ServiceComponent
        {
            private static string LogTag = "ClientService";
            private static int _seq = 0;
            private ComponentPort _port;

            public override override OnCreate()
            {
                Log.Info("OnCreate()");
                base.OnCreate();
                _port = new ComponentPort("ClientService");
                return true;
            }

            public override void OnStartCommand(AppControl appControl, bool restarted)
            {
                Log.Info("OnStartCommand()");
                base.OnStartCommand(appControl, restarted);
            }

            public override void OnDestroy()
            {
                Log.Info("OnDestroy()");
                base.OnDestroy();

                if (_port != null)
                {
                    _port.Dispose();
                    _port = null;
                }
            }
        }
    }
    ```

>     1. Create a component for instance for receiving component `ServerService.Tizen` as follows:

Author: Done. It's my mistake.

            {
                Log.Info("OnDestroy()");
                base.OnDestroy();

                if (_port != null)
                {
                    _port.RequestReceived -= OnRequestReceived;
                    _port.Dispose();
                    _port = null;
                }
            }
        }
    }
    ```

2.  Set up the receiving thread of the component.

    To have the receiving thread of the component wait for incoming requests, call the `WaitForEvent()` method of the `Tizen.Applications.ComponentBased.ComponentPort` class. 
    If the `WaitForEvent()` method is called in the main thread, the next line of calling the `WaitForEvent()` method will not be called until the `Cancel()` is called. 
    The `WaitForEvent()` method starts the main loop to wait for events from other components. To avoid blocking the main thread, it's recommended to use the `Tizen.Applications.ComponentBased.ComponentTask` class.
    To handle the access control, call `AddPrivilege()` of the `Tizen.Applications.ComponentBased.ComponentPort` class. If a client doesn't have permission, the request is rejected with the permission denied error.

>     To handle the received request, define and register an event handler for the `RequestReceived` event of the `Tizen.Applications.ComponentBased.ComponentPort` class as follows:

Author: Done

    ```
    {
        _port = new ComponentPort("ServerService");
        _port.RequestReceived += OnRequestReceived;
        _port.AddPrivilege("http://tizen.org/privilege/datasharing");
        _thread = new Thread(new ThreadStart(OnThread));
        _thread.Start();
    }

    private void OnThread()
    {
        Log.Info("OnTrhead()");
        _port.WaitForEvent();
    }

    private void OnRequestedReceived(object sender, RequestEventArgs args)
    {
        Log.Info("Sender: + args.Sender);
        var request = args.Request;
        if (request.GetType() == typeof(Request))
        {
            var req = (Request)request;
            Log.Info("Command: " + req.Command);
            Log.Info("Sequence: " + req.Sequence);
            Log.Info("Message: " + req.Message);

            if (args.IsReplyRequested)
            {
                args.Reply = new Response(req.Command, req.Sequence, 0);
            }
        }
    }
    ```

3.  Send the request in the sending component as follows:

>     1. Call `WaitForEvent()` of the `Tizen.Applications.ComponentBased.ComponentPort` class in the thread, if you want to receive the reply from the **ServerService**.

Author: Done

    2. Call `WaitForPort`() of the `Tizen.Applications.ComponentBased.ComponentPort` class to wait until the target port is ready.

    3. Use `Send()` of the `Tizen.Applications.ComponentBased.ComponentPort` class to send the request.

>     4. Provide the request to be sent as an instance of the `ClientService.Tizen.Request` class as follows:

Author: Done

    ```
    {
        _port = new ComponentPort("ClientService");
        _port.RequestReceived += OnRequestReceived;
        _thread = new Thread(new ThreadStart(OnThread));
        _thread.Start();
        ...
        await ComponentPort.WaitForPort("ServerService");

        try
        {
            var request = new Request("TestCommand", _seq++, "TestMessage");
            _port.Send("ServerService", 5000, request);
        }
        catch
        {
            Log.Error("Failed to send request");
        }
        ...
    }

    private void OnThread()
    {
        Log.Info("OnTrhead()");
        _port.WaitForEvent();
    }

    private void OnRequestedReceived(object sender, RequestEventArgs args)
    {
        Log.Info("Sender: + args.Sender);
        var response = args.Request;
        if (response.GetType() == typeof(Response))
        {
            var req = (Response)request;
            Log.Info("Command: " + req.Command);
            Log.Info("Sequence: " + req.Sequence);
            Log.Info("Result: " + req.Result);
        }
    }
    ```

<a name="task"></a>

> ## Use component task

Reviewer: Just for consistency across the pages suggesting this:
```suggestion
## Using component task

Author: Done
